core/src/main/scala/schema/provider/MetadataSchemaProvider.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.cassandra
+package schema.provider
+
+import cats.implicits._
+import cats.~>
+import com.datastax.driver.core._
+import freestyle._
+import freestyle.FreeS
+import freestyle.cassandra.schema.provider.metadata.SchemaConversions
+import freestyle.cassandra.schema.{SchemaDefinition, SchemaDefinitionProviderError}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.language.postfixOps
+
+class MetadataSchemaProvider(cluster: Cluster)
+    extends SchemaDefinitionProvider
+    with SchemaConversions {
+
+  def extractTables(keyspaceMetadata: KeyspaceMetadata): List[AbstractTableMetadata] =
+    keyspaceMetadata.getTables.asScala.toList
+
+  def extractIndexes(tableMetadataList: List[AbstractTableMetadata]): List[IndexMetadata] =
+    tableMetadataList.flatMap {
+      case (t: TableMetadata) => t.getIndexes.asScala.toList
+      case _                  => Nil
+    }
+
+  def extractUserTypes(keyspaceMetadata: KeyspaceMetadata): List[UserType] =
+    keyspaceMetadata.getUserTypes.asScala.toList
+
+  override def schemaDefinition: Either[SchemaDefinitionProviderError, SchemaDefinition] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    import freestyle.async.implicits._
+    import freestyle.cassandra.api._
+    import freestyle.cassandra.handlers.implicits._
+
+    implicit val clusterAPIInterpreter: ClusterAPI.Op ~> Future =
+      clusterAPIHandler[Future] andThen apiInterpreter[Future, Cluster](cluster)
+
+    def metadataF[F[_]](implicit clusterAPI: ClusterAPI[F]): FreeS[F, Metadata] =
+      for {
+        _        <- clusterAPI.connect
+        metadata <- clusterAPI.metadata
+        _        <- clusterAPI.close
+      } yield metadata
```

Comment:
There is this open PR in Cats : https://github.com/typelevel/cats/pull/1662 (issue https://github.com/typelevel/cats/issues/1630).

Doobie uses something similar in its [`Transactor`](https://github.com/tpolecat/doobie/blob/series/0.4.x/yax/core/src/main/scala/doobie/util/transactor.scala#L63). [`ConnectionIO`](https://github.com/tpolecat/doobie/blob/b10a679a0fa45ba68c3319022e04f2b0dac12fb2/yax/core/src/main/scala/doobie/free/connection.scala#L389) can contain side effects though (and can thus define a `MonadError` instance), and that is not something you can do with every `Free` / `FreeS`.

You can use the `guarantee` implementation from the PR linked to above after you interpret to an `M` with a `MonadError[M, Throwable]`
It is possible we can do something like they do in [quasar](https://github.com/quasar-analytics/quasar/blob/09a41912b28362f9f381d463d85cdcd5662d649d/effect/src/main/scala/quasar/effect/Failure.scala) to create a `MonadError` for `FreeS` if your algebras contain the `Error` effect, but that would need some more thought / work.